rfcs/global-naming-conventions.md

```diff
@@ -0,0 +1,58 @@
+- field names should use `camelCase`,
+- type names (including enum names) should use `PascalCase`, and
+- enum values should use `ALL_CAPS`.
+
+Many users have requested we provide support for these more "preferred" GraphQL naming conventions, especially `camelCase` field names [see issue [#3320](https://github.com/hasura/graphql-engine/issues/3320)].
+
+Users who care about using `camelCase` conventions currently have to use `custom_root_fields` for every field on every table.
+(One user has provided a [Python script](https://github.com/m-rgba/hasura-snake-to-camel/blob/main/snake_case_to_camelcase.py) which automatically does this by modifying the metadata.) Doing this manually is cumbersome and it seems there is demand for us to provide a setting for this at a higher level, e.g. for the entire GraphQL schema, or at the database or table level.
+
+Even if users can manually configure their own root fields, there are other generated names where the user has no control,
+e.g. parameter names such as `distinct_on` and `order_by`, nested field names such as `stddev_pop`, and type names such as `users_aggregate_fields`. A global setting to use `camelCase` should probably affect these names too.
+
+## Proposal - global `camelCase` option
+
+Add a global option to use `camelCase` for all field names and parameter names generated by Hasura, and to use `PascalCase` for all generated type names.
+We can automatically convert `snake_case` table and schema names to `camelCase` as appropriate.
+
+### Example
+
+For Postgres schema `my_schema` and table `user_accounts` we generate the following root fields:
+
+| Snake case (default)      | With `camelCase` option enabled |
+| --------------------      | ------------------------------- |
+| `my_schema_user_accounts` | `mySchemaUserAccounts` |
+| `my_schema_user_accounts_aggregate` | `mySchemaUserAccountsAggregate` |
+| `my_schema_user_accounts_by_pk` | `mySchemaUserAccountsByPk` |
+| `delete_my_schema_user_accounts` | `deleteMySchemaUserAccounts` |
+| `delete_my_schema_user_accounts_by_pk` | `deleteMySchemaUserAccountsByPk` |
+| `insert_my_schema_user_accounts` | `insertMySchemaUserAccounts` |
+| `insert_my_schema_user_accounts_one` | `insertMySchemaUserAccountsOne` |
+| `update_my_schema_user_accounts` | `updateMySchemaUserAccounts` |
+| `update_my_schema_user_accounts_by_pk` | `updateMySchemaUserAccountsByPk` |
+
+We would also convert all column names to `camelCase`.
+Similarly, all type names would use `PascalCase`, e.g. `my_schema_user_accounts_aggregate_fields` would become `MySchemaUserAccountsAggregateFields`.
+We would also modify other generated names e.g. `distinct_on` -> `distinctOn` and `stddev_pop` -> `stddevPop`.
+
+### Notes
+
+1. If the user specifies a `custom_name` for a table, we should use it as is, without trying to convert to `camelCase`. The user can use this to override the automated case conversion if there are edge cases where it doesn't do what they want.
```

Comment:
Currently, if the table has non-graphql compliant characters then it's discarded from the schema. We should continue to do that with the naming convention capability as well.